Added moderation capabilities

Dalethium opened the pull request:
Added the ability to approve or deny posts

Dalethium committed:
Added moderation capabilities

Dalethium commented:
@mikegcox hired me to code this, just thought I'd give him credit for getting it done :)

rhukster commented:
Very nice :).

flaviocopes commented:
Cool! Can you change tab indentation to 4 spaces, as per http://www.php-fig.org/psr/psr-2/? Also remove 

```
error_reporting(E_ALL);
 ini_set('display_errors', 1);
```

Dalethium committed:
fixed indentation (hopefully) and removed e_All

Dalethium commented:
I am used to K&R formatting, my bad. I removed E_all and such, is there anything else that needs done?

mikegcox commented:
Thanks for the additions to this code @Dalethium.  Hopefully the code can help the community.

flaviocopes commented on guestbook.php:
```diff
@@ -131,12 +335,27 @@ public function onTwigTemplatePaths()
     public function onDataTypeExcludeFromDataManagerPluginHook()
     {
-        $this->grav['admin']->dataTypesExcludedFromDataManagerPlugin[] = 'guestbook';
+        $this->grav['admin']->dataTypesExcludedFromDataManagerPlugin[] = 'a-guestbook';
```
Why this change?

mikegcox replied:
This was done by me so that the guestbook entries would show in the Data plugin in Admin.  I forgot to change it back before zipping my Grav folder and giving it to Dalethium.  It should not have been changed.

Dalethium replied:
should it be changed back?

flaviocopes commented on admin/templates/guestbook.html.twig:
```diff
@@ -0,0 +1,97 @@
+                    alert('Unexpected error');
+                });
+            });
+			$( document ).ready(function() {
```
Can you adjust the indentation here?

Dalethium replied:
Sure 1 sec

Dalethium committed:
fixed tabs to spaces

Dalethium committed:
changed data exclude hook

flaviocopes commented:
Thanks! I'll check the functionality tomorrow. Thanks @mikegcox too, we welcome this kind of contributions.

@Dalethium can you enable the option to allow edits from maintainers? https://github.com/blog/2247-improving-collaboration-with-forks so I can push directly to this PR if needed.

Dalethium commented:
I think it is on! I can directly add you as a contributor if you wish :)

flaviocopes commented:
Ok thanks!

flaviocopes commented:
Functionality-wise, this is going to force moderation. I think we have to add an option in the Plugin's settings and the default should be to not moderate, to preserve the current functionality. Also, I'd rather have a "your message is awaiting moderation" and not list any non-moderated comments, if moderation is on.

Also, currently I could not see the message after submit, but I see it being counted: 

![screen shot 2016-10-27 at 12 03 44](https://cloud.githubusercontent.com/assets/11783388/19763219/6911a10e-9c3d-11e6-8551-462bef79f31e.png)

and Admin needs some better UI, I see this

![screen shot 2016-10-27 at 12 05 50](https://cloud.githubusercontent.com/assets/11783388/19763282/a9a3c418-9c3d-11e6-91e4-f50d0d8ea86b.png)

It could take inspiration from Comments, which has this interface:

![screen shot 2016-10-27 at 12 05 44](https://cloud.githubusercontent.com/assets/11783388/19763299/b7646ea4-9c3d-11e6-8e94-1c580a6879a1.png)

Dalethium commented:
on it now

Dalethium committed:
UI improvements, and page buttons

Dalethium committed:
Php for previous commit

Dalethium commented:
hopefully that will meet your standards :)

Dalethium committed:
fixed errors that were added in last commit

Dalethium committed:
spacing

Dalethium committed:
php spacing

Dalethium committed:
minor change to UI

Dalethium commented:
What do you think of the page buttons? I can remove them if needed, I just figured it'd be marginally better than load more. The most recent will still be at the front, and they won't have to scroll through hell if they have hundreds of comments.

Side note: Approve all can be used to approve all comments from a outdated yaml format.

Dalethium commented:
the UI looks like this for me now:

![image](https://cloud.githubusercontent.com/assets/22735418/19782277/14cfb5f4-9c5b-11e6-8af2-bb441b3eb810.png)

Dalethium committed:
I'll get it right some day

mikegcox commented:
Here is my opinion... The pager is a nice addition but I am not sure it is necessary.  I think it would be seldom that a person would need to go through existing messages.  For moderation, all you would ever need to see is the latest entries that have not been approved.  If we keep the pager, the buttons should be made more elegant perhaps like Antimatter but there needs to be padding for sure.  

I also think the pager or load more button should be at the bottom.  It could be fixed to the bottom of the page so it does not float if floating is undesired.

The "Approve all Messages" button also needs padding.

Dalethium mentioned that front-end stuff is not his best skill-set so I don't mind taking a stab at it but I will wait until I get other opinions on the pagers and button styles.

Dalethium committed:
removed pager, changed design slightly

Dalethium committed:
added load all functionality

Dalethium commented:
I still have work to do, but I have made some changes at the request of @mikegcox  -

Dalethium committed:
removed redundant title and related js

Dalethium committed:
fixed a bug

The load all button would disappear when you clicked approve all

Dalethium committed:
Changed design, added option to go back

Dalethium committed:
added option

mikegcox commented:
![screen shot 2016-10-30 at 6 31 28 pm](https://cloud.githubusercontent.com/assets/5839231/19840572/2c85f142-9ecf-11e6-98a0-b4c225df4d88.jpg)

I did some work on formatting the guestbook admin page.  @flaviocopes, could you please let me know if you like the look of the screen-shot?  We used the same layout as the Data Manager plugin and I styled it using classes that I found on the dashboard.  I think this is very consistent with other Grav plugins.  It is responsive, the table layout used in the comments plugin is not so I like this option a lot more.

w00fz commented:
Very good job on this, it looks great!

flaviocopes commented:
Did you add some CSS that is not included? I see it differently: 
![screen shot 2016-10-31 at 13 29 32](https://cloud.githubusercontent.com/assets/11783388/19854198/14645dac-9f6e-11e6-9844-869d62875036.png)

flaviocopes commented:
BTW thanks for your effort in improving it :)

mikegcox commented:
He has not incorporated my styling changes yet.  I tried to fork it but didn't have permission.  We are working together on this and I keep asking for changes so he was getting tired yesterday.  I can add my changes, it is only to the twig file and language file.  I put a title at the top now too so it looks like this now:
![screen shot 2016-10-31 at 8 34 18 am](https://cloud.githubusercontent.com/assets/5839231/19854487/9721f80e-9f45-11e6-9285-843d7c91b491.jpg)

mikegcox commented:
This is the version with my styling changes.  He has an optional table layout in the code too but I asked hime to remove it as it is not necessary.  I also found a bug that needs to be fixed.  I also requested that all existing messages remain visible on the front-end upon upgrade but I have not tested that yet.

[Mike-changes-guestbook.zip](https://github.com/getgrav/grav-plugin-guestbook/files/561772/Mike-changes-guestbook.zip)

Dalethium committed:
fixed some bugs & @mikegcox updated UI

Dalethium committed:
added original functionality as a option

Dalethium committed:
added original functionality as a option

Dalethium commented:
Sorry for the late changes. I fixed a few bugs with removing deleted posts, I added original functionality as a config option, and I added the UI changes @mikegcox did with a small spacing change on the EM css

Dalethium commented:
![image](https://cloud.githubusercontent.com/assets/22735418/19895813/9158cdf6-a028-11e6-8a68-029a7dd08bc6.png)
This is what it looks like for me

Dalethium committed:
Changed functionality under guidance of @mikegcox

Dalethium committed:
Removed unsupported option

Dalethium committed:
Fixed so legacy clients can work as usual

Approves all messages on legacy message files

Dalethium committed:
added changed by @mikegcox

Dalethium committed:
@mikegcox updated the UI

Dalethium committed:
removed js alerts

mikegcox commented:
Thank you for your patience @Dalethium.  I think we have this working very well now.  It is now simplified and nicely styled.  
- When you first upgrade the plugin, the guestbook page of the website will not be affected at all. 
- When you go into admin, the exiting data file will be updated adding a guid and a moderated field for each entry and the site will remain un affected.
- when loaded in Admin, all un-moderated entries show up
- there is a button to show all moderated entries.
- When a delete button is pressed, the entry is removed 
- when the approve button is pressed the delete button goes away
- styling is accomplished using existing css classes from the Admin plugin except for `.center`.  That was added to a `<style>` block.
- the layout is responsive and looks very much like other plugins to Admin

I hope these changes can be incorporated into Guestbook.  The Guestbook plugin is a nice addition to Grav, but needs moderation capabilities.  Even my staging site was getting spammed.  We could tackle the Comments plugin next.

mikegcox commented:
![screen shot 2016-11-01 at 4 15 17 pm](https://cloud.githubusercontent.com/assets/5839231/19905766/7aa0e4ce-a04e-11e6-91d3-78d357625a6f.jpg)

The latest look

flaviocopes commented:
Thanks, I tested and it's good for me! I will merge _but_ I will add some code to make this an option, which means that moderation is optional (disabled by default) and that the Guestbook sidebar menu will only appear if it's enabled.

P.S. @mikegcox to avoid spam you can add a captcha to the guestbook / comment form

w00fz reviewed (approved):
I was going to suggest the optional factor. Other than that I'm approving as well!

mikegcox commented:
This is great news.  Thanks guys! 

@flaviocopes, I do have the captcha turned on and it works fine, but without the ability to moderate for the messages there are still undesirable messages being posted and they are live on the site by default.  So if someone has a beef with the site owner, they can voice it and it goes live.  If someone signs the guestbook and they have some spelling and grammar issues, a moderator can delete it from Admin and then repost from the website.  If it were just me moderating on my own site, this change would not have been a show-stopper, I could edit with cpanel or github or even FTP but for a non-technical person, it has to be moderated from the Admin plugin.

mikegcox commented:
@Dalethium, thanks for your patience... I know it was frustrating at times, but look what we can do now :)

flaviocopes commented:
@mikegcox yes, it is just a suggestion to avoid automated messages coming in 👍 

@Dalethium @mikegcox thanks for the contribution 😃

flaviocopes commented:
I just implemented moderation, if disabled in the plugin's settings, the admin panel disappears and  all messages are shown in the frontend.

I still get this issue which I need to look into, do you get the same? It's displaying  `Showing 3 messages of 3 (Note that some may be unmoderated)` but no messages appear, because moderation is enabled but no messages are approved.

flaviocopes commented:
Tracking in https://github.com/getgrav/grav-plugin-guestbook/issues/8
